californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java

```diff
@@ -196,10 +197,33 @@ public void receiveRequest(final Exchange exchange, final Request request) {
 
 		if (request.isDuplicate()) {
 			// Request is a duplicate, so resend ACK, RST or response
-			if (exchange.getCurrentResponse() != null) {
+			Response currentResponse = exchange.getCurrentResponse();
+			if (currentResponse != null) {
+				if (currentResponse.getType() == Type.NON || currentResponse.getType() == Type.CON) {
+					// separate response
+					if (request.isConfirmable()) {
+						// resend ACK,
+						// comply to RFC 7252, 4.2, cross-layer behavior
+						EmptyMessage ack = EmptyMessage.newACK(request);
+						sendEmptyMessage(exchange, ack);
+					}
+					if (currentResponse.isConfirmable()) {
+						// retransmission cycle
+						int failedCount;
+						synchronized (exchange) {
+							failedCount = exchange.getFailedTransmissionCount() + 1;
+							exchange.setFailedTransmissionCount(failedCount);
+						}
+						LOGGER.debug("Request Duplicate: retransmit response, failed: {}, response: {}",
+								new Object[] { failedCount, currentResponse });
+						currentResponse.retransmitting();
+						sendResponse(exchange, currentResponse);
```

Comment:
Using `sendResponse` instead of `lower().sendResponse()` is about preparing retransmission right ?

I asked my self if this could not go to some kind of race condition.
I mean we can face 2 preparing retransmission task at the same time. (because of classic CON response retransmission mechanism + reception of duplicate request)

Maybe not re-sending the response for CON response and just rely on the CON response retransmission mechanism is safer.

This is just an idea and if you think this is better like this. This PR sounds good to me.

Reply:
Even without that, some of my logs contains a failed retransmission counter with values much larger the 30 :-).